### PR TITLE
Add API to QgsMapLayerRegistry to rely less on ids

### PR DESCRIPTION
**Pull request by m-kuhn**

### CODE REVIEW

**Comment by mhugo** [comment]
Hi. Just to understand, in what cases would you want to rely less on ids ?

**Comment by m-kuhn** [comment]
I often manage lists of layers and want to remove layers from them when they disappear.

Before that would mean traverse it all

    for l in layerlist:
        if l.id() == layer_to_remove_id:
            layerlist.remove(l)

while now one can just do

    layerlist.remove(l)

Speaking more generally, I think layer ids are a good mean to serialize information (i.e. saving to projects) while at runtime their pointer/reference is just as unique and offers direct access to all its functionality and id's always need to be resolved via the registry, so I'd like to see the internal use of ids reduced as much as possible.

With this API change I'd like to make it easier to write code that does not rely on id's and to no longer *encourage* people to use the id wherever possible.

File: python/core/qgsmaplayerregistry.sip [hunk 1]
```diff
@@ -97,6 +97,21 @@ class QgsMapLayerRegistry : QObject
...
+     * Any canvases using the affected layers will need to remove them
+     *
+     * The layers being removed are deleted as well as the registry
+     * table entries.
+     *
+     * @param layers  The layers to remove
+     *
+     * @note As a side-effect QgsProject is made dirty.
+     */
+    void removeMapLayers( const QList<QgsMapLayer*>& theLayerIds );
```
Comment (3nids) [inline python/core/qgsmaplayerregistry.sip]: might be worth a rephrase of argument ;)

Comment (m-kuhn) [inline python/core/qgsmaplayerregistry.sip]: And I thought I might get away with it since you had your day off today ;)

**Comment by mhugo** [comment]
Makes sense.
Then what do you think about marking methods that access layers by id as obsolete ? and only keep mapLayer(id) -> maplayer* to convert from id to pointer ?

**Comment by m-kuhn** [comment]
Given that there may be scenarios where layer ids make sense and the (estimated) vast amount of usage of this API I refrained from it. Maybe a recommendation would be good.

**Pull request closed without merging**